File: frameworks/fw-wdio/specs/login.spec.js
describe('the-internet login (wdio via stage grid)', () => {
  it('logs in with valid credentials @smoke', async () => {
    await browser.url('https://the-internet.herokuapp.com/login');
    await $('#username').setValue('tomsmith');
    await $('#password').setValue('SuperSecretPassword!');
    await $('button[type="submit"]').click();
    await expect($('#flash')).toHaveTextContaining('You logged into a secure area!');
  });
  it('rejects invalid credentials @regression', async () => {
    await browser.url('https://the-internet.herokuapp.com/login');
    await $('#username').setValue('tomsmith');
    await $('#password').setValue('wrong-password');
    await $('button[type="submit"]').click();
    // explicit wait: #flash only exists after the navigation settles on the grid
    const flash = await $('#flash');
    await flash.waitForExist({ timeout: 15000 });
    await expect(flash).toHaveTextContaining('Your password is invalid!');
  });
});

File: frameworks/fw-ruby/spec/login_spec.rb
require "spec_helper"

RSpec.describe "the-internet login", type: :feature do
  it "navigates from home via link_text and logs in with valid credentials" do
    visit "/"
    click_link "Form Authentication" # link_text locator
    fill_in "username", with: "tomsmith"        # css #username
    fill_in "password", with: "SuperSecretPassword!"
    find(:xpath, "//button[@type='submit']").click # xpath locator
    selenium_wait { |d| d.find_elements(css: "#flash.success").any? }
    expect(page).to have_css("#flash.success", text: "You logged into a secure area!")
    expect(page).to have_xpath("//div[@id='content']//h2", text: "Secure Area")
  end

  it "rejects invalid credentials with an error flash" do
    visit "/login"
    fill_in "username", with: "not-a-user"
    fill_in "password", with: "wrong"
    click_button "Login"
    expect(page).to have_css("#flash.error", text: "Your username is invalid!")
  end

  it "logs out from the secure area" do
    visit "/login"
    fill_in "username", with: "tomsmith"
    fill_in "password", with: "SuperSecretPassword!"
    click_button "Login"
    expect(page).to have_css("a.button.secondary.radius", text: "Logout")
    click_link "Logout"
    expect(page).to have_css("#flash.success", text: "You logged out of the secure area!")
  end

  it "deliberate-failure: expects a marketing banner that does not exist" do
    visit "/login"
    # INTENTIONAL FAILURE — proves failure reporting works end to end.
    expect(page).to have_css("#marketing-banner", text: "Try KaneAI today!")
  end
end
